### PR TITLE
Scheduled run mode for bots

### PR DESCRIPTION
**Pull request by wagner-certat**
All bots with the parameter `"type": "oneshot"` in `runtime.conf` will stop after the first successful run. If the bot is disabled too, this is everything you need to schedule bots with cron.

See also the added docs for examples. Have fun :)

**Update**: Changed naming, see comments below.

### CODE REVIEW

File: docs/Bots.md [hunk 1]
```diff
@@ -5,13 +5,48 @@
...
 
-By default all of the bots are started when you start the whole botnet, however there is a possibility to 
-*disable* a bot. This means that the bot will not start every time you start the botnet, but you can start 
-and stop the bot if you specify the bot explicitly. To disable a bot, add the following to your 
-`runtime.conf`: `"enabled": false`. Be aware that this is **not** a normal parameter (like the others 
-described in this file). It is set outside of the `parameters` object in `runtime.conf`. Check the 
-[User-Guide](./User-Guide.md) for an example.
+## General remarks
 
+There are two different types of parameters, the initialization parameters are need to start the bot. The runtime parameters are needed by the bot itself during runtime.
```
Comment (aaronkaplan) [inline docs/Bots.md]: make a ":" after the first sentence.

**Comment by wagner-certat** [comment]
I reworded some sentences, replaced the example with a better one and added more explanations.

**Comment by SYNchroACK** [comment]
# Proposal

The following information is just a proposal to re-structure the running mode of bots, according to multiple questions and issues related to the need of different approaches to fetch feeds and other possible scenarios that you may think of.

**Note:** the names, values and other terms are merely names that I used to explain my proposal. They don't need to be called like that in the code or in the documentation, although, is extremely important to define clear and intuitive names in the implementation phase.

## Run Modes

Each bot has their own purposes, therefore there is a need for configuration that would allow diferent **run modes**. Looking at the current IntelMQ implementation and as far I'm aware, there are 2 run modes required as the following section shows:

### Configuration

| Run Mode  | Run Mode Config `parameter` | Extra Config  | Extra Config `parameter`  | Implement on Milestone |
|---|---|---|---|---|
| Scheduled | `"run_mode: "scheduled"` | Crontab Time String Syntax | `"crontab": "0 0,12 1 */2 *"` | v1.0 |
| Stream | `"run_mode: "stream"` | None | None | v1.0 |

### Description

**Run Mode: Scheduled** - bots that need to run on specific time (hour, day, etc), requires to be configured as `scheduled` with an extra parameter using the well-known crontab syntax.

**Run Mode: Stream** - bots that need to run indefinitely in time, requires to be configured as `stream`. 

### Examples
Scheduled `runtime.conf` configuration example:
```
{
  ...
  "run_mode": "scheduled",
  "crontab": "0 0,12 1 */2 *",
  ...
}
```

Stream `runtime.conf` configuration example:
```
{
  ...
  "run_mode": "stream",
  ...
}
```

### Bot Examples / Run Modes

| Bot / Type | Run Mode |
|---|---|
| Mail / Collector | Scheduled |
| HTTP / Collector | Scheduled |
| File / Output | Scheduled |
| HTTP-Stream / Collector | Stream |
| Abusech / Parser | Stream |
| Deduplicator / Expert | Stream |
| CymruWhois / Expert | Stream |

## Requirements

I think there are two requirements for this implementation, however, I would like to have your feedback on this @wagner-certat because of the `systemd` implementation. We need to evaluate if this proposal will be compliant with the possible `systemd` implementation. The requirements are the following:

* both run modes MUST keep using the `rate_limit` parameter.
* `intelmqctl` will be responsible to use the **python-crontab** library to write into crontab system (if run_mode is scheduled) or execute the bot (if run_mode is stream).






@aaronkaplan @wagner-certat @dmth give some feedback if you have chance. Thank you :)

**Comment by wagner-certat** [comment]
The updated PR now contains the changes from @SYNchroACK, renaming `oneshot` to `scheduled` and `type` to `run_mode`. Also, ctl now has subcommands called `enable` and `disable`.

**Comment by wagner-certat** [comment]
@ondj With the recent commits we are supporting multiple process managers. Could yout share your code for supervisord then?

File: docs/User-Guide.md [hunk 2]
```diff
@@ -19,6 +19,7 @@
       * [Web interface: IntelMQ Manager](#web-interface-intelmq-manager)
       * [Command-line interface: intelmqctl](#command-line-interface-intelmqctl)
         * [Botnet Concept](#botnet-concept)
+        * [Oneshot bots](#oneshot-bots)
```
Comment (aaronkaplan) [inline docs/User-Guide.md]: s/oneshot/scheduled/

File: docs/User-Guide.md [hunk 3]
```diff
@@ -123,7 +124,7 @@ cp -a examples/* .
 
 * `defaults.conf`: default values for bots and their behavior, e.g.
 error handling, log options and pipeline configuration. Will be removed in the [future](https://github.com/certtools/intelmq/issues/267).
-* `runtime.conf`: Configuration for the individual bots.
+* `runtime.conf`: Configuration for the individual bots. See Bots.md for more details.
```
Comment (aaronkaplan) [inline docs/User-Guide.md]: markdown link

File: docs/User-Guide.md [hunk 4]
```diff
@@ -278,7 +279,7 @@ This configuration is used by each bot to load the specific parameters associate
 }
 ```
 
-More examples can be found at `intelmq/etc/runtime.conf` directory in IntelMQ repository.
+More examples can be found at `intelmq/etc/runtime.conf` directory in IntelMQ repository. See Bots.md for more details.
```
Comment (aaronkaplan) [inline docs/User-Guide.md]: markdown link

File: docs/User-Guide.md [hunk 5]
```diff
@@ -453,7 +456,64 @@ Outputs are additionally logged to /opt/intelmq/var/log/intelmqctl
 
 #### Botnet Concept
 
-The botnet represents all currently configured bots. To get an overview which bots are running, use `intelmqctl status`or use IntelMQ Manager.
+The botnet represents all currently configured bots which are explicitly enabled. To get an overview which bots are running, use `intelmqctl status` or use IntelMQ Manager. Set `"enabled": true` in the runtime configuration to add a bot to botnet. By default, bots will be configured as `"enabled": false` therefore, does not belong to the botnet due protection reasons. See [Bots](Bots.md) for more details on configuration.
```
Comment (aaronkaplan) [inline docs/User-Guide.md]: grammar: 
" therefore, does not belong to the botnet due protection reasons" 

--> " therefore, such a bot does not belong to the botnet (by default)."

File: docs/User-Guide.md [hunk 6]
```diff
@@ -453,7 +456,64 @@ Outputs are additionally logged to /opt/intelmq/var/log/intelmqctl
...
+
+Disabled bots can still be started explicitly using `intelmqctl start <bot_id>`, but will keep marked as disabled if stopped. They are not started by `intelmqctl start` in analogy to the behavior of widely used initialization systems.
```
Comment (aaronkaplan) [inline docs/User-Guide.md]: "but will keep marked" --> "but will remain in the state `disabled` if stopped (and not be implicitly enabled by the `start` command).

File: docs/User-Guide.md [hunk 7]
```diff
@@ -453,7 +456,64 @@ Outputs are additionally logged to /opt/intelmq/var/log/intelmqctl
...
+
+
+#### Scheduled Run Mode
+
+In many cases it is useful to schedule a bot, for example, to collect information from a website every day at midnight. To do this, set `run_mode` to `scheduled` in the `runtime.conf` for the bot. Check the following example:
```
Comment (aaronkaplan) [inline docs/User-Guide.md]: "schedule a bot" -> "schedule a bot at a specific time (for example via cron(1)) ..."

File: docs/User-Guide.md [hunk 8]
```diff
@@ -453,7 +456,64 @@ Outputs are additionally logged to /opt/intelmq/var/log/intelmqctl
...
+    },
+},
+```
+
+You can schedule the bot with a crontab-entry like this:
+```
+0 0 * * * intelmqctl start blocklistde-apache-collector
+```
+
+Bots configured as scheduled will exit after the first successful run.
```
Comment (aaronkaplan) [inline docs/User-Guide.md]: ` scheduled `

File: docs/User-Guide.md [hunk 9]
```diff
@@ -453,7 +456,64 @@ Outputs are additionally logged to /opt/intelmq/var/log/intelmqctl
...
+
+
+#### Stream Run Mode
```
Comment (aaronkaplan) [inline docs/User-Guide.md]: continous

File: docs/intelmqctl.md [hunk 10]
```diff
@@ -210,13 +210,33 @@ The same as start and stop consecutively.
 ### reload
 The same as reload of every bot.
 
+### enable / disable
+The subcommands `enable` and `disable` are setting the according flags in the runtime.conf and thus make them part of the botnet.
```
Comment (aaronkaplan) [inline docs/intelmqctl.md]: "set the corresponding flags"
"in runtime.conf"

delete "and thus make them part of the botnet"

**Comment by wagner-certat** [comment]
@dmth should this be also mentioned in the NEWS.md file though it's not relevant for upgrading as we are backwards-compatible?

**Comment by dmth** [comment]
> @dmth should this be also mentioned in the NEWS.md file though it's not relevant for upgrading as we are backwards-compatible?

If you think this is a feature which should be propagated, the changelog would be a good place.

**Pull request merged**